pkg/template/jsonpath.go

```diff
@@ -0,0 +1,158 @@
+package template
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"regexp"
+
+	"k8s.io/client-go/third_party/forked/golang/template"
+	"k8s.io/client-go/util/jsonpath"
+)
+
+var (
+	// tektonVar is a regular expression for strings surrounded by $()
+	tektonVar = regexp.MustCompile(`^\$\((.*)\)$`)
+
+	// jsonRegexp is a regular expression for JSONPath expressions
+	// with or without the enclosing {} and the leading . inside the curly
+	// braces e.g.  'a.b' or '.a.b' or '{a.b}' or '{.a.b}
+	jsonRegexp = regexp.MustCompile(`^\{\.?([^{}]+)\}$|^\.?([^{}]+)$`)
+)
+
+// ParseJSONPath extracts a subset of the given JSON input
+// using the provided JSONPath expression.
+func ParseJSONPath(input interface{}, expr string) (string, error) {
+	j := jsonpath.New("").AllowMissingKeys(false)
+	buf := new(bytes.Buffer)
+
+	//First turn the expression into fully valid JSONPath
+	expr, err := TektonJSONPathExpression(expr)
+	if err != nil {
+		return "", err
+	}
+
+	if err := j.Parse(expr); err != nil {
+		return "", err
+	}
+
+	fullResults, err := j.FindResults(input)
+	if err != nil {
+		return "", err
+	}
+
+	for _, r := range fullResults {
+		if err := printResults(buf, r); err != nil {
+			return "", err
+		}
+	}
+
+	return buf.String(), nil
+}
+
+// PrintResults writes the results into writer
+// This is a slightly modified copy of the original
+// j.PrintResults from k8s.io/client-go/util/jsonpath/jsonpath.go
+// in that it uses calls `textValue()` for instead of `evalToText`
+// This is a workaround for kubernetes/kubernetes#16707
+func printResults(wr io.Writer, results []reflect.Value) error {
+	for i, r := range results {
+		text, err := textValue(r)
+		if err != nil {
+			return err
+		}
+		if i != len(results)-1 {
+			text = append(text, ' ')
+		}
+		if _, err := wr.Write(text); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// textValue translates reflect value to corresponding text
+// If the value if an array or map, it returns a JSON representation
+// of the value (as opposed to the internal go representation of the value)
+// Otherwise, the text value is from the `evalToText` function, originally from
+// k8s.io/client-go/util/jsonpath/jsonpath.go
```

Comment:
it seems like there are a few functions in this file that are copied directly from other places - maybe your commit message could explain a bit more about why? (im thinking it's b/c they were unexported and you need to use them?)

Reply:
added to commit message!